### DISCUSSION TITLE
docs: Add advanced mapResult examples

### DISCUSSION THREAD

**Discussion by alexandrebodin** [mikro-orm/mikro-orm#3477] [Q&A] [answered]
Hi there 👋 ,

Here is a simplified example of what I'm trying to achieve: 

- When using the repository to find with a nested `orderBy` I'm getting a single result even if the SQL is returning duplicates (due to left joins)
- When using the createQueryBuilder to make a similar query, I'm receiving all the duplicates (which seems normal) however I would have expected the mapResults to take action and de-duplicate the rows like the repository does. 

```js
// Articles have one Category <-> Category has many Articles

const categories = await categoryRepo.find(
    {},
    {
      orderBy: {
        articles: {
          name: "ASC",
        },
      },
    }
  );

// will return only one instance of each categories


const qb = orm.em.createQueryBuilder("Category");
const res = await qb
    .select("*")
    .orderBy({
      articles: {
        name: "ASC",
      },
    })
    .execute('all', true);

// returns multiple times the same categories
```

My question then is. Can we do the same map as what the repository does? if so It would be great to document it.

Sorry for the long message, hopefully that's clear :D

**Comment by B4nan** [comment]
You will need to be more specific here. `mapResult()` is a driver API, I consider that kinda internal, you should be using `em.map()`.

**Reply by alexandrebodin** [reply]
Yes I missed clicked an submitted I'm editting the message :D

**Comment by B4nan** [comment, marked as answer]
Using `qb.execute()` will always give you the raw results, you might want to use `qb.getResult()` instead, which handles mapping to entities (as well as merging them to the context) - that is equivalent to what EM/repository API does. 

```ts
const qb = orm.em.createQueryBuilder("Category");
const res = await qb
    .select("*")
    .orderBy({
      articles: {
        name: "ASC",
      },
    })
    .getResult();
```

This is documented in the very first section in the QB docs:

https://mikro-orm.io/docs/query-builder#executing-the-query

Awaiting the QB directly does the same. Also note that the `mapResults` parameter of `qb.execute()` defaults to `true`.

But the `qb.getResult()` (nor `em.mapResult()`) is not used for deduplication per-se, it handles mapping of joined results and "deduplicates" selected to-many relations, for that you would have to use `qb.joinAndSelect()` API. For your use case I believe adding distinct flag should help.

```ts
const qb = orm.em.createQueryBuilder("Category");
const res = await qb
    .select("*")
    .setFlag(QueryFlag.DISTINCT)
    .orderBy({
      articles: {
        name: "ASC",
      },
    })
    .getResult();
```

The deduplication you are talking about happens directly in entity manager, via simple `Utils.unique()` call (that works thanks to the entities being managed, so two entities with same PK will be identical by reference).

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/EntityManager.ts#L161

**Reply by alexandrebodin** [reply]
Yes I was curious to see the distinct isn't automatically being used when using implicit joins :). Thanks for the quick reply I missed the getResult(). 

The problem with using distinct sadly is that it will break with PostgreSQL and using an orderBy because you need to select the ordered column in your distinct too (which defeats the purpose) so you need to use either a sub query or a distinct ON. 

It's sadly not consistent across SQL databases. I guess it partially answers the 1st comment of not using distinct by default.

**Reply by B4nan** [reply]
> The problem with using distinct sadly is that it will break with PostgreSQL and using an orderBy because you need to select the ordered column in your distinct too (which defeats the purpose) so you need to use either a sub query or a distinct ON.

You are selecting all root entity properties and you would use the distinct flag on that, or no? 

> It's sadly not consistent across SQL databases. I guess it partially answers the 1st comment of not using distinct by default. 

Yeah distinct can be problematic, especially with auto-joined relations, not a universal tool, but here it felt like a good fit. 

Anyway, if you are fine with doing things the EM way, its basically this:

```ts
const qb = orm.em.createQueryBuilder("Category");
const res = await qb
    .select("*")
    .orderBy({
      articles: {
        name: "ASC",
      },
    });
const unique = Utils.unique(res);
```

**Reply by alexandrebodin** [reply]
> > The problem with using distinct sadly is that it will break with PostgreSQL and using an orderBy because you need to select the ordered column in your distinct too (which defeats the purpose) so you need to use either a sub query or a distinct ON.
> 
> You are selecting all root entity properties and you would use the distinct flag on that, or no?
> 
Yes but in PG you have to select the order by columns which breaks the distinct in that case. FYI https://blog.jooq.org/how-sql-distinct-and-order-by-are-related/
>
> > It's sadly not consistent across SQL databases. I guess it partially answers the 1st comment of not using distinct by default.
> 
> Yeah distinct can be problematic, especially with auto-joined relations, not a universal tool, but here it felt like a good fit.
> 
> Anyway, if you are fine with doing things the EM way, its basically this:
> 
> ```ts
> const qb = orm.em.createQueryBuilder("Category");
> const res = await qb
>     .select("*")
>     .orderBy({
>       articles: {
>         name: "ASC",
>       },
>     });
> const unique = Utils.unique(res);
> ```


Yes I have what I need :) Thank you.
